src/lib.rs

```diff
@@ -492,35 +546,76 @@ impl<L: SynthLanguage> Synthesizer<L> {
 
             layer.retain(|n| !n.all(|id| self.egraph[id].data.exact));
 
+            // no constants (if set)
+            log::info!("Made layer of {} nodes", layer.len());
             if iter > self.params.no_constants_above_iter {
-                let mut extract = Extractor::new(&self.egraph, NumberOfOps);
-
-                // maps ids to n_ops
-                let has_constants: HashSet<Id> = self
+                let constants: HashSet<Id> = self
                     .ids()
                     .filter_map(|id| {
-                        let (_cost, best) = extract.find_best(id);
-                        if best.as_ref().iter().any(|n| n.is_constant()) {
+                        let expr = &self.egraph[id].data.simplest;
+                        if expr.as_ref().iter().any(|n| n.is_constant()) {
                             Some(id)
                         } else {
                             None
                         }
                     })
                     .collect();
 
-                layer.retain(|n| n.all(|id| !has_constants.contains(&id)));
+                // layer.retain(|n| {
+                //     n.all(|id| !L::contains_ids_with_depth(&self.egraph, id, &constants, iter - 1))
+                // });
+                layer.retain(|n| n.all(|id| !constants.contains(&id)));
             }
+            
+            // deduplicate and filter bad constants
+            let mut cp = self.egraph.clone();
+            let mut max_id = cp.number_of_classes();
+            layer.retain(|node| {
+                let seen = &mut HashSet::<Id>::default();
+                if iter > self.params.ema_above_iter {
+                    let rec = node.to_recexpr(|id| self.egraph[id].data.simplest.as_ref());
+                    let rec2 = L::emt_generalize(&rec);
+                    let id = cp.add_expr(&rec2);
+                    if usize::from(id) < max_id {
```

Comment:
umm, what does this check do?

Reply:
Removes a newly enumerated node if it is redundant. Yes the egraph will automatically do this, but deduplicating nodes before cvec-matching and minimization is helpful since minimization takes a long time.

Reply:
huh that's very interesting! Cool!